0000-ldap-sync-tool.md

```diff
@@ -0,0 +1,139 @@
+- Feature Name: `mgr-ldapsync`: LDAP sync tool
+- Start Date: 2019, Oct 28
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The `mgr-ldapsync` is a tool that performs synchronisation of the
+users between Uyuni server and LDAP.
+
+# Motivation
+[motivation]: #motivation
+
+Uyuni server can authenticate users from LDAP via PAM. For this, users
+needs to be added to Uyuni with the flag "PAM authentication" and thus
+their credentials will be verified via LDAP directory. Synchronisation
+tool should transparently synchronise users from certain groups and/or
+roles in LDAP and reflect that in Uyuni server, overriding local
+changes.
+
+Often LDAP configuration at infrastructures is unpredictable. This
+tool should bring certain boundaries and over up most common LDAP
+setups for users and their roles management.
+
+# Detailed design
+[design]: #detailed-design
+
+The tool should work as an aggregator, collecting users from LDAP and
+Uyuni server synchronising them.
+
+## Functionality Aspects
+
+The following functions should be implemented:
+
+- Distinguish between local users in Uyuni and LDAP users.
+- Prevent UID clashes between the local users and LDAP users.
+- Prevent accidental modification or removal of the local user.
+- Flexible support various LDAP schemas, such as eDirectory, POSIX,
+  Active Directory etc.
+- Role management.
+- Dry run mode (preview only).
+
+## Configuration
+
+Configuration should be default in a separate file
+`/etc/rhn/ldapsync.conf`. Alternatively, user should be able to
+specify own path, in case the default one is not created or used.
+
+The main principle of the tool is to map specific roles and or groups
+in LDAP to the roles in the Uyuni server via the configuration. Adding
+or removing LDAP users from these groups or roles should reflect in
+Uyuni without additional manual changes in the Uyuni server.
+
+Configuration should be just a YAML file and the mapping should
+implement a relation between LDAP group to a set of roles from
+Uyuni. So each user in LDAP that is a member of such group would
+inherit these roles. For example:
+
+```yaml
+directory:
+  groups:
+    cn=sysop,ou=groups,dc=example,dc=com:
+      - activation_key_admin
+      - channel_admin
+      - config_admin
+```
+
+The configuration above would allow creating `groupOfNames` class in
+LDAP and add users via `member` attribute. As long as DN of that
+member is valid LDAP object and is not clashing with the local user in
+the Uyuni server, it should be synchronised. Similarly roles LDAP
+objects are configured as same as `groups` but via `roles`.
+
+## Frozen Users
+
+Frozen user is a local user in Uyuni, which has verified `org_admin`
+privileges (all) and is mentioned as "frozen" in the
+configuration. Such user is usually used as an emergency user, in case
+LDAP fails to respond and/or authenticate anyone. LDAP synchronisation
+should not proceed unless there is at least one "frozen" user.
```

Comment:
Does this mean if Uyuni Server loses connection to e. g. the Active Directory Domain Controller, then all the LDAP users (but the frozen ones) will be removed ? In the first try? Not even after some retries, timeouts or honoring AD credential cache time?

Looks dangerous to me.

Reply:
This is exactly the opposite: in case one adds to the LDAP a DN with the corresponding `cn` as same user as in Uyuni, say `cn=admin` and then syncs it and then removes that member from the DN group keeping the DN in the users class, then the Uyuni user will be removed permanently locking Uyuni. :wink:

This is exactly to make sure this never happens.

Reply:
I think we can detect a "connection refused" or "no route to host" error and in such cases we change nothing.
Only when we can connect to the LDAP server and a user configuration has changed or a user was removed we change things in Uyuni